discovery/gossiper.go

```diff
@@ -2343,70 +2311,227 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
+		// them.
+		pendingMsgs, err := d.cfg.MessageStore.MessagesForPeer(
+			compressedPeerPubKey,
+		)
+		if err != nil {
+			log.Errorf("Unable to retrieve pending messages for "+
+				"peer %x: %v", compressedPeerPubKey, err)
+			continue
+		}
+
+		// If there aren't any, there's no need for this handler to
+		// exist, so we can exit now.
+		if len(pendingMsgs) == 0 {
+			delete(d.peerConnectionHandlers, compressedPeerPubKey)
+			return
+		}
+
+		// Otherwise, we'll go ahead and send them. We'll make sure to
+		// check if the message is stale before sending the message to
+		// prevent sending it if so.
+		for _, msg := range pendingMsgs {
+			// Retrieve the short channel ID for which this message
+			// applies for logging purposes.
+			var shortChanID lnwire.ShortChannelID
+			switch msg := msg.(type) {
+			case *lnwire.AnnounceSignatures:
+				shortChanID = msg.ShortChannelID
+			case *lnwire.ChannelUpdate:
+				shortChanID = msg.ShortChannelID
+			default:
+				log.Warnf("Unexpected message %T found "+
+					"within store", msg)
+				continue
 			}
 
-			log.Errorf("unable to send AnnounceSignatures message "+
-				"to peer(%x): %v. Will retry when online.",
-				remotePeer.SerializeCompressed(), err)
+			if d.isStaleMsg(msg) {
+				log.Debugf("Removing stale %v message for "+
+					"channel=%v with peer=%x",
+					msg.MsgType(), shortChanID,
+					compressedPeerPubKey)
 
-			peerChan := make(chan lnpeer.Peer, 1)
-			d.cfg.NotifyWhenOnline(remotePeer, peerChan)
+				err := d.cfg.MessageStore.DeleteMessage(
+					msg, compressedPeerPubKey,
+				)
+				if err != nil {
+					log.Errorf("Unable to remove stale %v "+
+						"message for channel=%v with "+
+						"peer %x: %v", msg.MsgType(),
+						shortChanID,
+						compressedPeerPubKey, err)
+				}
 
-			select {
-			case <-peerChan:
-				// Retry sending.
-				log.Infof("Peer %x reconnected. Retry sending"+
-					" AnnounceSignatures.",
-					remotePeer.SerializeCompressed())
+				continue
+			}
 
-			case <-d.quit:
-				log.Infof("Gossiper shutting down, did not " +
-					"send AnnounceSignatures.")
-				return
+			// TODO(wilmer): is it worth checking if they're still
+			// online at this point?
+			if err := peer.SendMessage(false, msg); err != nil {
+				log.Errorf("Unable to send %v message for "+
+					"channel=%v to %x: %v", msg.MsgType(),
+					shortChanID, compressedPeerPubKey, err)
+				continue
 			}
+
+			log.Debugf("Successfully sent %v message for "+
+				"channel=%v with peer=%x upon reconnection",
+				msg.MsgType(), shortChanID,
+				compressedPeerPubKey)
 		}
 
-		log.Infof("Sent channel announcement proof to remote peer: %x",
-			remotePeer.SerializeCompressed())
-	}()
+		// Once the messages are sent, we can request a notification for
+		// when the peer disconnects. We do this to prevent the loop
+		// from proceeding and avoid repeatedly requesting online
+		// notifications for the peer.
+		offlineChan := d.cfg.NotifyWhenOffline(compressedPeerPubKey)
```

Comment:
nice, this is cool :)

Reply:
on second thought, it might be good to register for this notification as soon as the peer's online notification is received.

Let's say the peer comes online and we begin trying to send messsages, if the peer disconnects and then reconnects before this is called, we will wait until this new connection is done before attempting to send again. Moving this up will reduce the likelihood of that happening